File: app/db/FlowTable.ts
import { readFileSync, writeFileSync, existsSync } from 'node:fs'
import { join } from 'node:path'

// ========== 数据模型 ==========

/** 前端 @xyflow/react 节点 */
export interface FlowNode {
  id: string
  type: string
  position: { x: number; y: number }
  data: Record<string, unknown>
  parentId?: string
  extent?: 'parent'
  style?: Record<string, unknown>
}

/** 前端 @xyflow/react 连线 */
export interface FlowEdge {
  id: string
  source: string
  target: string
  animated?: boolean
  style?: Record<string, unknown>
}

/** 一条编排流程 */
export interface Flow {
  id: string
  name: string
  description: string
  nodes: FlowNode[]
  edges: FlowEdge[]
  createdAt: string
  updatedAt: string
}

// ========== 数据库路径 ==========
const DEFAULT_DB_PATH = join(process.cwd(), 'app', 'db', 'flow.json')

// ========== FlowTable 操作类 ==========

export class FlowTable {
  private dbPath: string

  constructor(dbPath?: string) {
    this.dbPath = dbPath ?? DEFAULT_DB_PATH
    this.ensureDB()
  }

  // ---------- 私有：文件 I/O ----------

  /** 确保 flow.json 存在 */
  private ensureDB(): void {
    if (!existsSync(this.dbPath)) {
      writeFileSync(this.dbPath, '[]', 'utf-8')
    }
  }

  /** 读取全部数据 */
  private readDB(): Flow[] {
    try {
      const raw = readFileSync(this.dbPath, 'utf-8')
      return JSON.parse(raw) as Flow[]
    } catch {
      return []
    }
  }

  /** 写入全部数据 */
  private writeDB(data: Flow[]): void {
    writeFileSync(this.dbPath, JSON.stringify(data, null, 2), 'utf-8')
  }

  // ---------- 公开：CRUD ----------

  /** 获取所有流程 */
  getAll(): Flow[] {
    return this.readDB()
  }

  /** 按 ID 获取单个流程 */
  getById(id: string): Flow | undefined {
    return this.readDB().find((f) => f.id === id)
  }

  /** 按名称模糊搜索 */
  search(keyword: string): Flow[] {
    const kw = keyword.toLowerCase()
    return this.readDB().filter(
      (f) =>
        f.name.toLowerCase().includes(kw) ||
        (f.description?.toLowerCase().includes(kw) ?? false),
    )
  }

  /** 新建流程（自动生成 ID 和时间戳） */
  create(data: Omit<Flow, 'id' | 'createdAt' | 'updatedAt'>): Flow {
    const flows = this.readDB()
    const now = new Date().toISOString()
    const flow: Flow = {
      id: this.generateId(),
      ...data,
      createdAt: now,
      updatedAt: now,
    }
    flows.push(flow)
    this.writeDB(flows)
    return flow
  }

  /** 更新流程 */
  update(id: string, data: Partial<Omit<Flow, 'id' | 'createdAt'>>): Flow | null {
    const flows = this.readDB()
    const index = flows.findIndex((f) => f.id === id)
    if (index === -1) return null

    const existing = flows[index]!
    const updated: Flow = {
      id: existing.id,
      name: data.name ?? existing.name,
      description: data.description ?? existing.description,
      nodes: data.nodes ?? existing.nodes,
      edges: data.edges ?? existing.edges,
      createdAt: existing.createdAt,
      updatedAt: new Date().toISOString(),
    }
    flows[index] = updated
    this.writeDB(flows)
    return updated
  }

  /** 保存流程（存在则更新，不存在则新建） */
  save(flow: Flow): Flow {
    const existing = this.getById(flow.id)
    if (existing) {
      const result = this.update(flow.id, flow)
      return result!
    }
    return this.create(flow)
  }

  /** 仅保存 nodes 和 edges（轻量更新） */
  setFlow(id: string, nodes: FlowNode[], edges: FlowEdge[]): Flow | null {
    return this.update(id, { nodes, edges })
  }

  /** 删除流程 */
  delete(id: string): boolean {
    const flows = this.readDB()
    const filtered = flows.filter((f) => f.id !== id)
    if (filtered.length === flows.length) return false
    this.writeDB(filtered)
    return true
  }

  /** 获取流程总数 */
  count(): number {
    return this.readDB().length
  }

  // ---------- 工具 ----------

  /** 生成唯一 ID */
  private generateId(): string {
    return `flow_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`
  }
}

// ========== 单例导出 ==========
export const flowTable = new FlowTable()

File: db/flow_table.py
"""Flow 数据持久化 —— 基于 JSON 文件的 CRUD"""

import json
import os
import time
import random
import string
from datetime import datetime, timezone

from flow import Flow, FlowEdge, FlowNode

DEFAULT_DB_PATH = os.path.join(os.path.dirname(__file__), "flow.json")


def _to_flow(item: dict) -> Flow:
    """将 dict 转为 Flow 数据类"""
    nodes = [
        FlowNode(
            id=n["id"],
            type=n["type"],
            position=n.get("position", {}),
            data=n.get("data", {}),
            parentId=n.get("parentId"),
            extent=n.get("extent"),
            style=n.get("style", {}),
        )
        for n in item.get("nodes", [])
    ]
    edges = [
        FlowEdge(
            id=e["id"],
            source=e["source"],
            target=e["target"],
            animated=e.get("animated", False),
            style=e.get("style", {}),
        )
        for e in item.get("edges", [])
    ]
    return Flow(
        id=item["id"],
        name=item["name"],
        description=item.get("description", ""),
        nodes=nodes,
        edges=edges,
        createdAt=item.get("createdAt", ""),
        updatedAt=item.get("updatedAt", ""),
    )


def _from_flow(flow: Flow) -> dict:
    """将 Flow 数据类转为 dict"""
    return {
        "id": flow.id,
        "name": flow.name,
        "description": flow.description,
        "nodes": [
            {
                "id": n.id,
                "type": n.type,
                "position": n.position,
                "data": n.data,
                **({"parentId": n.parentId} if n.parentId else {}),
                **({"extent": n.extent} if n.extent else {}),
                **({"style": n.style} if n.style else {}),
            }
            for n in flow.nodes
        ],
        "edges": [
            {
                "id": e.id,
                "source": e.source,
                "target": e.target,
                **({"animated": e.animated} if e.animated else {}),
                **({"style": e.style} if e.style else {}),
            }
            for e in flow.edges
        ],
        "createdAt": flow.createdAt,
        "updatedAt": flow.updatedAt,
    }


class FlowTable:
    """基于 JSON 文件的 Flow 数据表"""

    def __init__(self, db_path: str | None = None):
        self.db_path = db_path or DEFAULT_DB_PATH
        self._ensure_db()

    # ---------- 文件 I/O ----------

    def _ensure_db(self) -> None:
        if not os.path.exists(self.db_path):
            with open(self.db_path, "w", encoding="utf-8") as f:
                json.dump([], f)

    def _read_db(self) -> list[dict]:
        try:
            with open(self.db_path, "r", encoding="utf-8") as f:
                return json.load(f)
        except (json.JSONDecodeError, FileNotFoundError):
            return []

    def _write_db(self, data: list[dict]) -> None:
        with open(self.db_path, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=2)

    # ---------- CRUD ----------

    def get_all(self) -> list[Flow]:
        return [_to_flow(item) for item in self._read_db()]

    def get_all_dicts(self) -> list[dict]:
        return self._read_db()

    def get_by_id(self, flow_id: str) -> Flow | None:
        for item in self._read_db():
            if item["id"] == flow_id:
                return _to_flow(item)
        return None

    def search(self, keyword: str) -> list[Flow]:
        kw = keyword.lower()
        return [
            _to_flow(item)
            for item in self._read_db()
            if kw in item.get("name", "").lower()
            or kw in item.get("description", "").lower()
        ]

    def create(self, data: dict) -> Flow:
        flows = self._read_db()
        now = datetime.now(timezone.utc).isoformat()
        flow = {
            "id": self._generate_id(),
            "name": data.get("name", ""),
            "description": data.get("description", ""),
            "nodes": data.get("nodes", []),
            "edges": data.get("edges", []),
            "createdAt": now,
            "updatedAt": now,
        }
        flows.append(flow)
        self._write_db(flows)
        return _to_flow(flow)

    def update(self, flow_id: str, data: dict) -> Flow | None:
        flows = self._read_db()
        for i, item in enumerate(flows):
            if item["id"] == flow_id:
                updated = {
                    "id": item["id"],
                    "name": data.get("name", item["name"]),
                    "description": data.get("description", item.get("description", "")),
                    "nodes": data.get("nodes", item["nodes"]),
                    "edges": data.get("edges", item["edges"]),
                    "createdAt": item.get("createdAt", ""),
                    "updatedAt": datetime.now(timezone.utc).isoformat(),
                }
                flows[i] = updated
                self._write_db(flows)
                return _to_flow(updated)
        return None

    def save(self, flow: Flow) -> Flow:
        existing = self.get_by_id(flow.id)
        if existing:
            result = self.update(flow.id, _from_flow(flow))
            if result:
                return result
        return self.create(_from_flow(flow))

    def set_flow(
        self, flow_id: str, nodes: list[dict], edges: list[dict]
    ) -> Flow | None:
        return self.update(flow_id, {"nodes": nodes, "edges": edges})

    def delete(self, flow_id: str) -> bool:
        flows = self._read_db()
        filtered = [f for f in flows if f["id"] != flow_id]
        if len(filtered) == len(flows):
            return False
        self._write_db(filtered)
        return True

    def count(self) -> int:
        return len(self._read_db())

    # ---------- 工具 ----------

    @staticmethod
    def _generate_id() -> str:
        suffix = "".join(random.choices(string.ascii_lowercase + string.digits, k=6))
        return f"flow_{int(time.time() * 1000)}_{suffix}"


# 单例
flow_table = FlowTable()
